sasmodels/kernelcuda.py

```diff
@@ -412,55 +418,58 @@ class GpuKernel(Kernel):
-    *dtype* is the kernel precision
+    *model* is the GpuModel object to call
 
-    The resulting call method takes the *pars*, a list of values for
-    the fixed parameters to the kernel, and *pd_pars*, a list of (value,weight)
-    vectors for the polydisperse parameters.  *cutoff* determines the
-    integration limits: any points with combined weight less than *cutoff*
-    will not be calculated.
+    The kernel is derived from :class:`Kernel`, providing the
+    :meth:`call_kernel` method to evaluate the kernel for a given set of
+    parameters.  Because of the need to move the q values to the GPU before
+    evaluation, the kernel is instantiated for a particular set of q vectors,
+    and can be called many times without transfering q each time.
 
     Call :meth:`release` when done with the kernel instance.
     """
-    def __init__(self, kernel, dtype, model_info, q_vectors):
-        # type: (cl.Kernel, np.dtype, ModelInfo, List[np.ndarray]) -> None
+    #: SAS model information structure
+    info = None # type: ModelInfo
+    #: kernel precision
+    dtype = None # type: np.dtype
+    #: kernel dimensions (1d or 2d)
+    dim = "" # type: str
+    #: calculation results, updated after each call to :meth:`_call_kernel`
+    result = None # type: np.ndarray
+
+    def __init__(self, model, q_vectors):
+        # type: (GpuModel, List[np.ndarray]) -> None
+        dtype = model.dtype
         self.q_input = GpuInput(q_vectors, dtype)
-        self.kernel = kernel
+        self._model = model
         # F16 isn't sufficient, so don't support it
         self._as_dtype = np.float64 if dtype == generate.F64 else np.float32
 
         # attributes accessed from the outside
         self.dim = '2d' if self.q_input.is_2d else '1d'
-        self.info = model_info
-        self.dtype = dtype
+        self.info = model.info
+        self.dtype = model.dtype
 
         # holding place for the returned value
         nout = 2 if self.info.have_Fq and self.dim == '1d' else 1
         extra_q = 4  # total weight, form volume, shell volume and R_eff
         self.result = np.empty(self.q_input.nq*nout+extra_q, dtype)
 
-        # Inputs and outputs for each kernel call
-        # Note: res may be shorter than res_b if global_size != nq
+        # allocate result value on gpu
         width = ((self.result.size+31)//32)*32 * self.dtype.itemsize
-        self.result_b = cuda.mem_alloc(width)
-        self._need_release = [self.result_b]
+        self._result_b = cuda.mem_alloc(width)
 
     def _call_kernel(self, call_details, values, cutoff, magnetic, effective_radius_type):
         # type: (CallDetails, np.ndarray, np.ndarray, float, bool) -> np.ndarray
         # Arrange data transfer to card
         details_b = cuda.to_device(call_details.buffer)
         values_b = cuda.to_device(values)
 
-        kernel = self.kernel[1 if magnetic else 0]
-        args = [
+        name = 'Iq' if self.dim == '1d' else 'Imagnetic' if magnetic else 'Iqxy'
```

Comment:
We might want to spread this across multiple lines.  The nested inline conditionals are unidiomatic enough to be confusing (e.g. is iT "Iq' if self.dim == '1d' else ('Imagnetic' if magnetic else 'Iqxy')" or "(Iq' if self.dim == '1d' else 'Imagnetic') if magnetic else 'Iqxy'").  Also, it does push the line just past the pep8 80 column recommendation.  I have been letting that one slide, but it's just an extra excuse to turn this into an if-elif-else setup.

Reply:
It evaluates left-to-right, so I find that it reads pretty cleanly:

```value = val1 if cond1 else val2 if cond2 else ... else default```

Also, this case is exactly 80 columns.